# **Stock Sentiment Analysis Using News Headlines**

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/content/drive/MyDrive/Stock-Sentiment-Analysis-Using-News-Headlines-main/data/Combined_News_DJIA.csv')

In [3]:
df.head()
#label 0 = stock price has negative impact(stock price is going to decrease next day  )
#label 1 = stock price has positive impact(stock price is going to increase )

Date  Label                                               Top1  \
0  2008-08-08      0  b"Georgia 'downs two Russian warplanes' as cou...   
1  2008-08-11      1  b'Why wont America and Nato help us? If they w...   
2  2008-08-12      0  b'Remember that adorable 9-year-old who sang a...   
3  2008-08-13      0  b' U.S. refuses Israel weapons to attack Iran:...   
4  2008-08-14      1  b'All the experts admit that we should legalis...   

                                                Top2  \
0            b'BREAKING: Musharraf to be impeached.'   
1        b'Bush puts foot down on Georgian conflict'   
2                 b"Russia 'ends Georgia operation'"   
3  b"When the president ordered to attack Tskhinv...   
4  b'War in South Osetia - 89 pictures made by a ...   

                                                Top3  \
0  b'Russia Today: Columns of troops roll into So...   
1  b"Jewish Georgian minister: Thanks to Israeli ...   
2  b'"If we had no sexual harassment we would hav...   
3  b' Israel clears troops who killed Reuters cam...   
4  b'Swedish wrestler Ara Abrahamian throws away ...   

                                                Top4  \
0  b'Russian tanks are moving towards the capital...   
1  b'Georgian army flees in disarray as Russians ...   
2  b"Al-Qa'eda is losing support in Iraq because ...   
3  b'Britain\'s policy of being tough on drugs is...   
4  b'Russia exaggerated the death toll in South O...   

                                                Top5  \
0  b"Afghan children raped with 'impunity,' U.N. ...   
1      b"Olympic opening ceremony fireworks 'faked'"   
2  b'Ceasefire in Georgia: Putin Outmaneuvers the...   
3  b'Body of 14 year old found in trunk; Latest (...   
4  b'Missile That Killed 9 Inside Pakistan May Ha...   

                                                Top6  \
0  b'150 Russian tanks have entered South Ossetia...   
1  b'What were the Mossad with fraudulent New Zea...   
2  b'Why Microsoft and Intel tried to kill the XO...   
3  b'China has moved 10 *million* quake survivors...   
4  b"Rushdie Condemns Random House's Refusal to P...   

                                                Top7  \
0  b"Breaking: Georgia invades South Ossetia, Rus...   
1  b'Russia angered by Israeli military sale to G...   
2  b'Stratfor: The Russo-Georgian War and the Bal...   
3  b"Bush announces Operation Get All Up In Russi...   
4  b'Poland and US agree to missle defense deal. ...   

                                                Top8  ...  \
0  b"The 'enemy combatent' trials are nothing but...  ...   
1  b'An American citizen living in S.Ossetia blam...  ...   
2  b"I'm Trying to Get a Sense of This Whole Geor...  ...   
3             b'Russian forces sink Georgian ships '  ...   
4  b'Will the Russians conquer Tblisi? Bet on it,...  ...   

                                               Top16  \
0  b'Georgia Invades South Ossetia - if Russia ge...   
1  b'Israel and the US behind the Georgian aggres...   
2  b'U.S. troops still in Georgia (did you know t...   
3                      b'Elephants extinct by 2020?'   
4  b'Bank analyst forecast Georgian crisis 2 days...   

                                               Top17  \
0                b'Al-Qaeda Faces Islamist Backlash'   
1  b'"Do not believe TV, neither Russian nor Geor...   
2       b'Why Russias response to Georgia was right'   
3  b'US humanitarian missions soon in Georgia - i...   
4  b"Georgia confict could set back Russia's US r...   

                                               Top18  \
0  b'Condoleezza Rice: "The US would not act to p...   
1  b'Riots are still going on in Montreal (Canada...   
2  b'Gorbachev accuses U.S. of making a "serious ...   
3             b"Georgia's DDOS came from US sources"   
4  b'War in the Caucasus is as much the product o...   

                                               Top19  \
0  b'This is a busy day:  The European Union has ...   
1    b'China to overtake US as largest manufac

In [4]:
train = df[df['Date'] < '20150101']
test = df[df['Date'] > '20141231']

In [5]:
# Removing punctuations
data=train.iloc[:,2:27]
data.replace("[^a-zA-Z]"," ",regex=True, inplace=True)

# Renaming column names for ease of access
list1= [i for i in range(25)]
new_Index=[str(i) for i in list1]
data.columns= new_Index
data.head(5)

0  \
0  b Georgia  downs two Russian warplanes  as cou...   
1  b Why wont America and Nato help us  If they w...   
2  b Remember that adorable   year old who sang a...   
3  b  U S  refuses Israel weapons to attack Iran ...   
4  b All the experts admit that we should legalis...   

                                                   1  \
0            b BREAKING  Musharraf to be impeached     
1        b Bush puts foot down on Georgian conflict    
2                 b Russia  ends Georgia operation     
3  b When the president ordered to attack Tskhinv...   
4  b War in South Osetia      pictures made by a ...   

                                                   2  \
0  b Russia Today  Columns of troops roll into So...   
1  b Jewish Georgian minister  Thanks to Israeli ...   
2  b  If we had no sexual harassment we would hav...   
3  b  Israel clears troops who killed Reuters cam...   
4  b Swedish wrestler Ara Abrahamian throws away ...   

                                                   3  \
0  b Russian tanks are moving towards the capital...   
1  b Georgian army flees in disarray as Russians ...   
2  b Al Qa eda is losing support in Iraq because ...   
3  b Britain  s policy of being tough on drugs is...   
4  b Russia exaggerated the death toll in South O...   

                                                   4  \
0  b Afghan children raped with  impunity   U N  ...   
1      b Olympic opening ceremony fireworks  faked     
2  b Ceasefire in Georgia  Putin Outmaneuvers the...   
3  b Body of    year old found in trunk  Latest  ...   
4  b Missile That Killed   Inside Pakistan May Ha...   

                                                   5  \
0  b     Russian tanks have entered South Ossetia...   
1  b What were the Mossad with fraudulent New Zea...   
2  b Why Microsoft and Intel tried to kill the XO...   
3  b China has moved     million  quake survivors...   
4  b Rushdie Condemns Random House s Refusal to P...   

                                                   6  \
0  b Breaking  Georgia invades South Ossetia  Rus...   
1  b Russia angered by Israeli military sale to G...   
2  b Stratfor  The Russo Georgian War and the Bal...   
3  b Bush announces Operation Get All Up In Russi...   
4  b Poland and US agree to missle defense deal  ...   

                                                   7  \
0  b The  enemy combatent  trials are nothing but...   
1  b An American citizen living in S Ossetia blam...   
2  b I m Trying to Get a Sense of This Whole Geor...   
3             b Russian forces sink Georgian ships     
4  b Will the Russians conquer Tblisi  Bet on it ...   

                                                   8  \
0  b Georgian troops retreat from S  Osettain cap...   
1  b Welcome To World War IV  Now In High Definit...   
2  b The US military was surprised by the timing ...   
3  b The commander of a Navy air reconnaissance s...   
4  b Russia exaggerating South Ossetian death tol...   

                                                   9  ...  \
0  b Did the U S  Prep Georgia for War with Russia    ...   
1  b Georgia s move  a mistake of monumental prop...  ...   
2    b U S  Beats War Drum as Iran Dumps the Dollar   ...   
3  b     of CNN readers  Russia s actions in Geor...  ...   
4  b  Musharraf expected to resign rather than fa...  ...   

                                                  15  \
0  b Georgia Invades South Ossetia   if Russia ge...   
1  b Israel and the US behind the Georgian aggres...   
2  b U S  troops still in Georgia  did you know t...   
3                      b Elephants extinct by          
4  b Bank analyst forecast Georgian crisis   days...   

                                                  16  \
0                b Al Qaeda Faces Islamist Backlash    
1  b  Do not believe TV  neither Russian nor Geor...   
2       b Why Russias response to Georgia was right    
3  b US humanitarian missions soon in Georgia   i...   
4  b Georgia confict could set bac

In [6]:
# Convertng headlines to lower case
for index in new_Index:
    data[index]=data[index].str.lower()
data.head(1)

0  \
0  b georgia  downs two russian warplanes  as cou...   

                                         1  \
0  b breaking  musharraf to be impeached     

                                                   2  \
0  b russia today  columns of troops roll into so...   

                                                   3  \
0  b russian tanks are moving towards the capital...   

                                                   4  \
0  b afghan children raped with  impunity   u n  ...   

                                                   5  \
0  b     russian tanks have entered south ossetia...   

                                                   6  \
0  b breaking  georgia invades south ossetia  rus...   

                                                   7  \
0  b the  enemy combatent  trials are nothing but...   

                                                   8  \
0  b georgian troops retreat from s  osettain cap...   

                                                   9  ...  \
0  b did the u s  prep georgia for war with russia    ...   

                                                  15  \
0  b georgia invades south ossetia   if russia ge...   

                                    16  \
0  b al qaeda faces islamist backlash    

                                                  17  \
0  b condoleezza rice   the us would not act to p...   

                                                  18  \
0  b this is a busy day   the european union has ...   

                                                  19  \
0  b georgia will withdraw       soldiers from ir...   

                                                  20  \
0  b why the pentagon thinks attacking iran is a ...   

                                                  21  \
0  b caucasus in crisis  georgia invades south os...   

                                                  22  \
0  b indian shoe manufactory    and again in a se...   

                                                  23  \
0  b visitors suffering from mental illnesses ban...   

                                         24  
0  b no help for mexico s kidnapping surge   

[1 rows x 25 columns]

In [7]:
' '.join(str(x) for x in data.iloc[1,0:25])

'b why wont america and nato help us  if they wont help us now  why did we help them in iraq   b bush puts foot down on georgian conflict  b jewish georgian minister  thanks to israeli training  we re fending off russia   b georgian army flees in disarray as russians advance   gori abandoned to russia without a shot fired  b olympic opening ceremony fireworks  faked   b what were the mossad with fraudulent new zealand passports doing in iraq   b russia angered by israeli military sale to georgia  b an american citizen living in s ossetia blames u s  and georgian leaders for the genocide of innocent people  b welcome to world war iv  now in high definition   b georgia s move  a mistake of monumental proportions   b russia presses deeper into georgia  u s  says regime change is goal  b abhinav bindra wins first ever individual olympic gold medal for india  b  u s  ship heads for arctic to define territory  b drivers in a jerusalem taxi station threaten to quit rather than work for their 

In [8]:
headlines = []
for row in range(0,len(data.index)):
    headlines.append(' '.join(str(x) for x in data.iloc[row,0:25]))

In [9]:
headlines[0]

'b georgia  downs two russian warplanes  as countries move to brink of war  b breaking  musharraf to be impeached   b russia today  columns of troops roll into south ossetia  footage from fighting  youtube   b russian tanks are moving towards the capital of south ossetia  which has reportedly been completely destroyed by georgian artillery fire  b afghan children raped with  impunity   u n  official says   this is sick  a three year old was raped and they do nothing  b     russian tanks have entered south ossetia whilst georgia shoots down two russian jets   b breaking  georgia invades south ossetia  russia warned it would intervene on so s side  b the  enemy combatent  trials are nothing but a sham  salim haman has been sentenced to       years  but will be kept longer anyway just because they feel like it   b georgian troops retreat from s  osettain capital  presumably leaving several hundred people killed   video   b did the u s  prep georgia for war with russia   b rice gives green

## **1.Accuracy Using Count Vectorizer**

In [10]:
###CountVectorizer(converts sentences into vectors)
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.ensemble import RandomForestClassifier

## **Example:**

**the cat is sitting on the mat** 

**the dog is eating** 

**the cat is dancing** 

**combination of the sentences :** 

**the cat dog is sitting eating dancing on the mat** 

**vector form of first sentence : 1 1 0 1 1 0 0 1 1 1** 

**ngram combine words like in this code it will combine first two words((the cat)(the dog))** 

In [11]:
## implement BAG OF WORDS
countvector=CountVectorizer(ngram_range=(2,2))
traindataset=countvector.fit_transform(headlines)

## **RandomForest Classifier**

In [12]:
# implement RandomForest Classifier
randomclassifier=RandomForestClassifier(n_estimators=200,criterion='entropy')
randomclassifier.fit(traindataset,train['Label'])

RandomForestClassifier(criterion='entropy', n_estimators=200)

In [13]:
## Predict for the Test Dataset
test_transform= []
for row in range(0,len(test.index)):
    test_transform.append(' '.join(str(x) for x in test.iloc[row,2:27]))
test_dataset = countvector.transform(test_transform)
predictions = randomclassifier.predict(test_dataset)

In [14]:
test.loc[1611,:]

Date                                            2015-01-02
Label                                                    1
Top1     Most cases of cancer are the result of sheer b...
Top2     Iran dismissed United States efforts to fight ...
Top3     Poll: One in 8 Germans would join anti-Muslim ...
Top4     UK royal family's Prince Andrew named in US la...
Top5     Some 40 asylum-seekers refused to leave the bu...
Top6     Pakistani boat blows self up after India navy ...
Top7     Sweden hit by third mosque arson attack in a week
Top8            940 cars set alight during French New Year
Top9     Salaries for top CEOs rose twice as fast as av...
Top10    Norway violated equal-pay law, judge says: Jud...
Top11    Imam wants radical recruiters of Muslim youth ...
Top12    Saudi Arabia beheaded 83 people in 2014, the m...
Top13    'A living hell' for slaves on remote South Kor...
Top14    Worlds 400 richest get richer, adding $92bn in...
Top15    Rental Car Stereos Infringe Copyright, Music R.

In [15]:
predictions

array([1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1,
       1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0,
       1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1,
       0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,

In [16]:
## Import library to check accuracy
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

In [17]:
matrix=confusion_matrix(test['Label'],predictions)
print('Confusion Matrix is: ')
print(matrix)

score=accuracy_score(test['Label'],predictions)
print('Accuracy: ',score)

report=classification_report(test['Label'],predictions)
print('Precision Report: ')
print(report)

Confusion Matrix is: 
[[132  54]
 [  2 190]]
Accuracy:  0.8518518518518519
Precision Report: 
              precision    recall  f1-score   support

           0       0.99      0.71      0.83       186
           1       0.78      0.99      0.87       192

    accuracy                           0.85       378
   macro avg       0.88      0.85      0.85       378
weighted avg       0.88      0.85      0.85       378



##**2.Accuracy Using Tfidf Vectorizer** 

In [18]:
###TfidfVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier

In [19]:
## implement BAG OF WORDS
tfidfvector=TfidfVectorizer(ngram_range=(2,2))
traindataset=tfidfvector.fit_transform(headlines)

## **A.Random Forest classifier**

In [20]:
# implement RandomForest Classifier
randomclassifier=RandomForestClassifier(n_estimators=200,criterion='entropy')
randomclassifier.fit(traindataset,train['Label'])

RandomForestClassifier(criterion='entropy', n_estimators=200)

In [21]:
## Predict for the Test Dataset
test_transform= []
for row in range(0,len(test.index)):
    test_transform.append(' '.join(str(x) for x in test.iloc[row,2:27]))
test_dataset = tfidfvector.transform(test_transform)
predictions = randomclassifier.predict(test_dataset)

In [22]:
predictions

array([1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1,
       1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0,
       1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1,
       0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,
       1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0,

In [24]:
## Import library to check accuracy
from sklearn.metrics import classification_report,confusion_matrix,accuracy_score

In [25]:
matrix=confusion_matrix(test['Label'],predictions)
print('Confusion Matrix is: ')
print(matrix)

score=accuracy_score(test['Label'],predictions)
print('Accuracy: ',score)

report=classification_report(test['Label'],predictions)
print('Precision Report: ')
print(report)

Confusion Matrix is: 
[[149  37]
 [ 25 167]]
Accuracy:  0.8359788359788359
Precision Report: 
              precision    recall  f1-score   support

           0       0.86      0.80      0.83       186
           1       0.82      0.87      0.84       192

    accuracy                           0.84       378
   macro avg       0.84      0.84      0.84       378
weighted avg       0.84      0.84      0.84       378



## **B. Naive Bayes Classifier(MultinomialNB)**

In [26]:
from sklearn.naive_bayes import MultinomialNB

In [27]:
naive=MultinomialNB()
naive.fit(traindataset,train['Label'])

MultinomialNB()

In [28]:
## Predict for the Test Dataset
test_transform= []
for row in range(0,len(test.index)):
    test_transform.append(' '.join(str(x) for x in test.iloc[row,2:27]))
test_dataset = tfidfvector.transform(test_transform)
predictions = naive.predict(test_dataset)

In [29]:
predictions

array([1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1,
       1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0,
       1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1,
       0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,

In [30]:
matrix=confusion_matrix(test['Label'],predictions)
print('Confusion Matrix is: ')
print(matrix)

score=accuracy_score(test['Label'],predictions)
print('Accuracy: ',score)

report=classification_report(test['Label'],predictions)
print('Precision Report: ')
print(report)

Confusion Matrix is: 
[[130  56]
 [  0 192]]
Accuracy:  0.8518518518518519
Precision Report: 
              precision    recall  f1-score   support

           0       1.00      0.70      0.82       186
           1       0.77      1.00      0.87       192

    accuracy                           0.85       378
   macro avg       0.89      0.85      0.85       378
weighted avg       0.89      0.85      0.85       378



# **Accuracy:**

**1.Using CountVectorizer and RandomForest: 85.44%**

**2.Using TfidfVectorizer and RandomForest: 83%**

**3.Using TfidfVectorizer and Multinomial Naive Bayes: 85.18%**